### DISCUSSION TITLE
close quarto preview

### DISCUSSION THREAD

**Discussion by jcolomb** [quarto-dev/quarto-cli#3718] [Q&A]
Hi,

I have a specific qmd file that should be re-run when the data is updated. I would like to use `quarto preview` to preview it, but once it is run once, it will run the specific qmd file only if the file itself is modified.

With blowdown, I was using the`blogdown::stop_server()` command to reinitialise the preview, but there seem to be no equivalent in quarto. Am I missing something or should I put this into the  ideas section. (note this was not ideal, as some cache was still there and the preview would still work even when some catastrophic changes were made).

I would like a `quarto preview -reboot` kind of command that would forget about the previous preview commands.

**Comment by jcolomb** [comment]
Note that deleting the content of the _site folder does not do the trick, the code is not run again.

**Reply by cscheid** [reply]
Can you share a minimal example that's causing this issue so we can test it on our side?

**Reply by jcolomb** [reply]
difficult, may get to it next year. I think any .qmd file having some code will re-run only if the file is changed (i.e. not if the data read in that code chunk changes) ? I would like a way to force running all file (as it happens the first time you run `quarto preview`)

**Reply by cscheid** [reply]
I guess what I don't understand is what you want `quarto preview -reboot` to achieve. If you're rerunning a _new_ instance of `quarto preview`, why doesn't killing the old `quarto preview` process and restarting work? We use that ourselves on a daily basis.

**Comment by jcolomb** [comment]
How do you kill the old process?

 Blogdown had a function for that, could not find the equivalent for quarto... 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Reply by cscheid** [reply]
`Ctrl-C` on the terminal should do it.

**Reply by danmackinlay** [reply]
are there other ways to kill it than `ctrl-c`? I find it remarkably hard to kill the deno server process [automatically](https://gist.github.com/danmackinlay/03a9c50b6719391aaee9291c9615f429), and I'm not sure why.
I would like to [restart the server at intervals](https://danmackinlay.name/notebook/quarto_websites.html#restart-server).

**Reply by cscheid** [reply]
You should be able to use your shell job controls (`kill`, `killall`, etc) to take more drastic actions. The reason I wouldn't recommend doing that automatically is that you're bound to hit a situation where you `kill -9` a process and it will leave temporary state strewn about.

Crash recovery is quite hard to do well, and, honestly, Quarto doesn't do well at all on that.

**Reply by danmackinlay** [reply]
What would I kill? I find it quite hard to acquire a PID in the obvious way
Currently I use `killall deno` but that kills other deno processes (And other quarto processes)

**Reply by cscheid** [reply]
Reading your script now, I see `kill $pid` there. That would have been my first guess. Other ideas:

- `kill -9 $pid` 
- (I only know how to do this in Linux with `awk '{print $5}' < /proc/$pid/stat`) Obtain the process group for $pid, set it to $pgid, and then `kill -9 -- -${pgid}` (which I believe would send SIGKILL to all processes with group id `$pgid`)
